### PR TITLE
fix: derive files from tsconfigs which extend

### LINKED ISSUE #3999: [labs/analyzer]: No source files found in package when tsconfig extends
### Which package(s) are affected?

Other/unknown (please mention in description)

### Description

Analyzer can't find source files when `extends` is used in tsconfig.

```
Analyzer error: No source files found in package.
original error
  Error: No source files found in package.
      at Analyzer.getPackage (file:///home/bennyp/Developer/patternfly/patternfly-elements/node_modules/@lit-labs/analyzer/lib/analyzer.js:39:19)
      at analyze (file:///home/bennyp/Developer/patternfly/patternfly-elements/tools/pfe-tools/cli/commands/analyze.js:9:28)
```

### Reproduction

Given https://github.com/patternfly/patternfly-elements/blob/367d186b15e2e92487bcf596364cfda06526411e/elements/tsconfig.json

and using this wrapper function:

```ts
async function analyze(argv: Opts) {
  const { packagePath } = argv;
  const path = packagePath instanceof URL ? fileURLToPath(packagePath) : join(process.cwd(), packagePath);
  const analyzer = createPackageAnalyzer(path as AbsolutePath);
  try {
    const p = analyzer.getPackage();
    console.log(p);
  } catch (e: any) {
    console.log(`Analyzer error: ${e.message}`);
    console.group(`original error`);
    console.error(e);
    console.groupEnd();
  }
}
```

and in the root of https://github.com/patternfly/patternfly-elements, running 

```js
await analyze({ packagePath: 'elements/tsconfig.json' });
```

### Workaround

replace the extends clause with the adjusted contents of tsconfig.settings.json. I will also need to replace `includes` with `files: []`, as the analyzer  **also does not pick up on files from `includes`**

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

0.8.0

### Browser/OS/Node environment

fedora 38

### PR DESCRIPTION
**Pull request by bennypowers**
Fixes #3999

### CODE REVIEW

**Comment by bennypowers** [comment]
`parseJsonConfigFileContent` seems to be resolving the includes/excludes/files just fine, but does not pick up on the project references or `*.css` declaration file, and it in fact looks like it's including everything in the pfe monorepo. Could be i'm holding it wrong, but other ts tooling has been working ok.

File: packages/labs/analyzer/src/lib/analyze-package.ts [hunk 1]
```diff
@@ -41,15 +41,16 @@ export const createPackageAnalyzer = (
   let commandLine: ts.ParsedCommandLine;
   if (ts.sys.fileExists(configFileName)) {
     const configFile = ts.readConfigFile(configFileName, ts.sys.readFile);
+    const existingOptions: CompilerOptions = {};
     if (options.exclude !== undefined) {
-      (configFile.config.exclude ??= []).push(...options.exclude);
+      existingOptions.exclude = options.exclude;
```
Comment (43081j) [inline packages/labs/analyzer/src/lib/analyze-package.ts:46]: Do you know for sure that this doesn't override the excludes in the config file? Maybe you should still concat the two?

File: packages/labs/analyzer/src/lib/analyze-package.ts [hunk 2]
```diff
@@ -41,15 +41,16 @@ export const createPackageAnalyzer = (
...
     }
     commandLine = ts.parseJsonConfigFileContent(
       configFile.config /* json */,
-      ts.sys /* host */,
-      packagePath /* basePath */,
-      undefined /* existingOptions */,
-      path.relative(packagePath, configFileName) /* configFileName */
+      ts.sys,
```
Comment (43081j) [inline packages/labs/analyzer/src/lib/analyze-package.ts:50]: Would be nice to keep the comments I think 👀

File: packages/labs/analyzer/src/lib/analyze-package.ts (hunk 1 above)
Comment (bennypowers) [reply packages/labs/analyzer/src/lib/analyze-package.ts:46]: i don't know for sure, but internally, that param is called `existingOptions`, and it is processed in https://github.com/microsoft/TypeScript/blob/9701f55f7217d7bd56d28e73b250444efcefa8dd/src/compiler/commandLineParser.ts#L2859

**Comment by bennypowers** [comment]
as of [a82b0e7](https://github.com/lit/lit/pull/4006/commits/a82b0e7862fa3a77065f9574b01ec2c898b962d5), the analyzer picks up on my source files, but it does not seem to parse the tsconfig correctly. namely, I get a bunch of errors which i don't normally get with typescript, probably related to build mode

<details>
<summary>errors</summary>

```

 core/pfe-core/controllers/cascade-controller.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/functions/debounce.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/controllers/cascade-controller.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/functions/debounce.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/controllers/internals-controller.ts
  TS2339 Property 'role' does not exist on type 'ARIAMixin'.

 core/pfe-core/controllers/internals-controller.ts
  TS2339 Property 'ariaInvalid' does not exist on type 'ARIAMixin'.

 core/pfe-core/controllers/internals-controller.ts
  TS2339 Property 'validity' does not exist on type 'ElementInternals'.

 core/pfe-core/controllers/internals-controller.ts
  TS2339 Property 'setValidity' does not exist on type 'ElementInternals'.

 core/pfe-core/controllers/internals-controller.ts
  TS2339 Property 'setValidity' does not exist on type 'ElementInternals'.

 core/pfe-core/controllers/internals-controller.ts
  TS2339 Property 'checkValidity' does not exist on type 'ElementInternals'.

 core/pfe-core/controllers/internals-controller.ts
  TS2339 Property 'checkValidity' does not exist on type 'ElementInternals'.

 core/pfe-core/controllers/internals-controller.ts
  TS2339 Property 'reportValidity' does not exist on type 'ElementInternals'.

 core/pfe-core/controllers/internals-controller.ts
  TS2339 Property 'reportValidity' does not exist on type 'ElementInternals'.

 core/pfe-core/controllers/overflow-controller.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/functions/isElementInView.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/controllers/overflow-controller.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/functions/isElementInView.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/decorators.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/bound.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/decorators.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/bound.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/decorators.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/cascades.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/decorators.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/cascades.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/decorators.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/deprecation.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/decorators.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/deprecation.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/decorators.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/initializer.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/decorators.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/initializer.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/decorators.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/observed.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/decorators.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/observed.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/decorators.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/time.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/decorators.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/time.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/decorators.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/trace.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/decorators.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/decorators/trace.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/decorators/cascades.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/controllers/cascade-controller.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/decorators/cascades.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/controllers/cascade-controller.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all files or use an 'include' pattern.

 core/pfe-core/decorators/initializer.ts
  TS6059 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/controllers/light-dom-controller.ts' is not under 'rootDir' '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. 'rootDir' is expected to contain all source files.

 core/pfe-core/decorators/initializer.ts
  TS6307 File '/home/bennyp/Developer/patternfly/patternfly-elements/core/pfe-core/controllers/light-dom-controller.ts' is not listed within the file list of project '/home/bennyp/Developer/patternfly/patternfly-elements/elements/tsconfig.json'. Projects must list all 
```

</details>

**Comment by bennypowers** [comment]
By tweaking my monorepo project's tsconfigs i reduced the number of errors to 17, all of which appear to be `lib` problems, presumably because analyzer uses typescript 4.7.4 whereas I'm up to 5.1.6 by now

```
core/pfe-core/controllers/internals-controller.ts:8:27 - error TS2339: Property 'role' does not exist on type 'ARIAMixin'.

8   declare role: ARIAMixin['role'];
                            ~~~~~~
```

<details><summary>rest of the errors</summary>

```
core/pfe-core/controllers/internals-controller.ts:22:34 - error TS2339: Property 'ariaInvalid' does not exist on type 'ARIAMixin'.

22   declare ariaInvalid: ARIAMixin['ariaInvalid'];
                                    ~~~~~~~~~~~~~
core/pfe-core/controllers/internals-controller.ts:65:28 - error TS2339: Property 'validity' does not exist on type 'ElementInternals'.

65     return this.#internals.validity;
                              ~~~~~~~~
core/pfe-core/controllers/internals-controller.ts:108:52 - error TS2339: Property 'setValidity' does not exist on type 'ElementInternals'.

108   setValidity(...args: Parameters<ElementInternals['setValidity']>) {
                                                       ~~~~~~~~~~~~~
core/pfe-core/controllers/internals-controller.ts:109:28 - error TS2339: Property 'setValidity' does not exist on type 'ElementInternals'.

109     return this.#internals.setValidity(...args);
                               ~~~~~~~~~~~
core/pfe-core/controllers/internals-controller.ts:112:54 - error TS2339: Property 'checkValidity' does not exist on type 'ElementInternals'.

112   checkValidity(...args: Parameters<ElementInternals['checkValidity']>) {
                                                         ~~~~~~~~~~~~~~~
core/pfe-core/controllers/internals-controller.ts:113:28 - error TS2339: Property 'checkValidity' does not exist on type 'ElementInternals'.

113     return this.#internals.checkValidity(...args);
                               ~~~~~~~~~~~~~
core/pfe-core/controllers/internals-controller.ts:116:55 - error TS2339: Property 'reportValidity' does not exist on type 'ElementInternals'.

116   reportValidity(...args: Parameters<ElementInternals['reportValidity']>) {
                                                          ~~~~~~~~~~~~~~~~
core/pfe-core/controllers/internals-controller.ts:117:28 - error TS2339: Property 'reportValidity' does not exist on type 'ElementInternals'.

117     return this.#internals.reportValidity(...args);
                               ~~~~~~~~~~~~~~
elements/pf-jump-links/pf-jump-links-item.ts:35:5 - error TS2345: Argument of type '{ role: string; }' is not assignable to parameter of type 'Partial<ARIAMixin>'.
  Object literal may only specify known properties, and 'role' does not exist in type 'Partial<ARIAMixin>'.

35     role: 'listitem'
       ~~~~~~~~~~~~~~~~
elements/pf-popover/pf-popover.ts:193:5 - error TS2304: Cannot find name 'satisfies'.

193   } satisfies Record<AlertSeverity, string>) as [AlertSeverity, string][]);
        ~~~~~~~~~
elements/pf-popover/pf-popover.ts:193:5 - error TS2554: Expected 1 arguments, but got 3.

193   } satisfies Record<AlertSeverity, string>) as [AlertSeverity, string][]);
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
elements/pf-popover/pf-popover.ts:193:15 - error TS2693: 'Record' only refers to a type, but is being used as a value here.

193   } satisfies Record<AlertSeverity, string>) as [AlertSeverity, string][]);
                  ~~~~~~
elements/pf-progress-stepper/pf-progress-step.ts:55:5 - error TS2345: Argument of type '{ role: string; }' is not assignable to parameter of type 'Partial<ARIAMixin>'.
  Object literal may only specify known properties, and 'role' does not exist in type 'Partial<ARIAMixin>'.

55     role: 'listitem',
       ~~~~~~~~~~~~~~~~
elements/pf-progress-stepper/pf-progress-stepper.ts:38:5 - error TS2345: Argument of type '{ role: string; ariaValueNow: string; }' is not assignable to parameter of type 'Partial<ARIAMixin>'.
  Object literal may only specify known properties, and 'role' does not exist in type 'Partial<ARIAMixin>'.

38     role: 'progressbar',
       ~~~~~~~~~~~~~~~~~~~
elements/pf-tabs/BaseTab.ts:43:21 - error TS2339: Property 'role' does not exist on type 'ElementInternals'.

43     this.#internals.role = 'tab';
                       ~~~~
elements/pf-tabs/BaseTabPanel.ts:22:21 - error TS2339: Property 'role' does not exist on type 'ElementInternals'.

22     this.#internals.role = 'tabpanel';
                       ~~~~

17 errors

```

</details>

So with that in mind i think this is ready for review

**Comment by justinfagnani** [comment]
Tagging in @kevinpschaaf and @rictic both for the review, but also... we need to think about supporting multiple versions of TypeScript.

**Comment by justinfagnani** [comment]
I think that is because the analyzer is actually trying to find that file on disk. It should really give up if it can't.

**Comment by bennypowers** [comment]
here's my calling code:

```ts
async function analyze(argv: Opts) {
  const { packagePath } = argv;
  const path = packagePath instanceof URL ? fileURLToPath(packagePath) : join(process.cwd(), packagePath);
  const analyzer = createPackageAnalyzer(path as AbsolutePath, {
    exclude: [
      '**/*.spec.ts',
      '**/*.e2e.ts',
      '**/*.d.ts',
      '**/*.js',
    ],
  });
  try {
    const p = analyzer.getPackage();
    console.log('got package', p);
  } catch (e: any) {
    if (Array.isArray(e.diagnostics)) {
      console.log(e.diagnostics.at(0));
      const formattedDiagnostics = formatDiagnostics(e.diagnostics);
      console.log(`${formattedDiagnostics}\n${e.diagnostics.length} errors`);
    } else {
      console.group(`Analyzer error`);
      console.error(e);
      console.groupEnd();
    }
  }
}
```


here is the diagnostic object from `e.diagnostic`, redacted for readability, where you can see that the code i get from analyzer is 2323:

```ts
{
  file: <ref *1> SourceFileObject { ... },
  start: 566,
  length: 21,
  category: 1,
  code: 2323,
  messageText: 'Could not resolve specifier ./BaseAccordion.css to filesystem path.'
}
```

**Comment by 43081j** [comment]
all good @bennypowers , it turns out Peter already fixed it in #3866! non-issue 🥳

File: packages/labs/analyzer/src/lib/analyze-package.ts [hunk 3]
```diff
@@ -47,9 +47,9 @@ export const createPackageAnalyzer = (
     commandLine = ts.parseJsonConfigFileContent(
       configFile.config /* json */,
       ts.sys /* host */,
-      packagePath /* basePath */,
-      undefined /* existingOptions */,
-      path.relative(packagePath, configFileName) /* configFileName */
+      isDirectory ? packagePath : path.dirname(packagePath) /* basePath */,
+      {} /* existingOptions */,
+      configFileName /* configFileName */
```
Comment (kevinpschaaf) [inline packages/labs/analyzer/src/lib/analyze-package.ts:52]: So, this whole API is really funky, but note early on I had left this comment:

https://github.com/lit/lit/blob/9a7b6546c286a964cafb707812353f33c6f0113c/packages/labs/analyzer/src/lib/analyzer.ts#L56-L60

I believe this empirical finding was why the `configFileName` _argument_ here was being calculated relative to `basePath`. But I guess if you're not getting the errors I was having with `packagePath` show up both in `basePath` and `configFileName` and the tests are passing, I probably just misjudged what was going on...?

**Review by kevinpschaaf — APPROVED** [review]
Ok, with these changes I've confirmed I can generate a CEM from [MWC](https://github.com/material-components/material-web)[*] source, which besides the monorepo tests, was the other thing I had been testing analyzer against, so while I can't really explain why the comment I linked from the older code would seem to indicate this change would break something, it doesn't seem to. 

So LGTM, and if we find out this breaks something we'll just need to add more tests.

*To get a successful run on MWC, I did need to upgrade the TS version the analyzer was using to match what MWC was using (per known issue in https://github.com/lit/lit/issues/4018), and then skip over a couple of unresolvable symbols in MWC code added since I last tested (needs more investigation, but wasn't the failure mode this type of change would lead to).

**Review by justinfagnani — APPROVED** [review]

**Pull request merged**